core/src/main/java/com/linecorp/armeria/client/HttpClientDelegate.java

```diff
@@ -215,24 +230,58 @@ private void acquireConnectionAndExecute0(ClientRequestContext ctx, Endpoint end
         }
     }
 
-    private ProxyConfig getProxyConfig(SessionProtocol protocol, Endpoint endpoint) {
+    private void resolveProxyConfig(SessionProtocol protocol, Endpoint endpoint, ClientRequestContext ctx,
+                                    BiConsumer<@Nullable ProxyConfig, @Nullable Throwable> onComplete) {
         final ProxyConfig proxyConfig = factory.proxyConfigSelector().select(protocol, endpoint);
         requireNonNull(proxyConfig, "proxyConfig");
 
+        final ServiceRequestContext serviceCtx = ServiceRequestContext.currentOrNull();
+        final ProxiedAddresses capturedProxiedAddresses = serviceCtx == null ? null
+                                                          : serviceCtx.proxiedAddresses();
+
+        // DirectProxyConfig does not have proxyAddress as its field.
+        if (proxyConfig.proxyAddress() != null) {
+            final Future<InetSocketAddress> resolveFuture = addressResolverGroup
+                    .getResolver(ctx.eventLoop().withoutContext())
+                    .resolve(proxyConfig.proxyAddress());
+
+            resolveFuture.addListener(future -> {
+                if (future.isSuccess()) {
+                    final InetSocketAddress resolvedAddress = (InetSocketAddress) future.getNow();
+                    if (resolvedAddress != null && !resolvedAddress.isUnresolved()) {
+                        final ProxyConfig newProxyConfig = proxyConfig.withNewProxyAddress(resolvedAddress);
+                        onComplete.accept(maybeHAProxy(newProxyConfig, capturedProxiedAddresses), null);
+                    } else {
+                        logger.warn("Resolved address is invalid or unresolved: {}. " +
+                                    "Using the previous address instead.", resolvedAddress);
+                        onComplete.accept(maybeHAProxy(proxyConfig, capturedProxiedAddresses), null);
+                    }
+                } else {
+                    final Throwable cause = future.cause();
+                    logger.warn("Failed to refresh {}'s ip address. " +
+                                "Using the previous address instead. reason: {}",
+                                proxyConfig.proxyAddress(),
+                                cause != null ? cause.getMessage() : "Unknown Error");
+                    onComplete.accept(maybeHAProxy(proxyConfig, capturedProxiedAddresses), null);
```

Comment:
Question) Shouldn't the cause be exposed to the user?
Also, if this is done I don't think we need to log dns resolution failure since the user can access the throwable directly

```suggestion
                    onComplete.accept(maybeHAProxy(proxyConfig, capturedProxiedAddresses), cause);
```